kernel/cost.h

```diff
@@ -111,6 +111,223 @@ struct CellCosts
+	std::set<RTLIL::IdString> unknown_cell_area;
+
+	statdata_t operator+(const statdata_t &other) const
+	{
+		statdata_t sum = other;
+	#define X(_name) sum._name += _name;
+		STAT_NUMERIC_MEMBERS
+	#undef X
+		for (auto &it : num_cells_by_type)
+			sum.num_cells_by_type[it.first] += it.second;
+		return sum;
+	}
+
+	statdata_t operator*(int other) const
+	{
+		statdata_t sum = *this;
+	#define X(_name) sum._name *= other;
+		STAT_NUMERIC_MEMBERS
+	#undef X
+		for (auto &it : sum.num_cells_by_type)
+			it.second *= other;
+		return sum;
+	}
+
+	statdata_t()
+	{
+	#define X(_name) _name = 0;
+		STAT_NUMERIC_MEMBERS
+	#undef X
+	}
+
+	statdata_t(RTLIL::Design *design, RTLIL::Module *mod, bool width_mode, const dict<IdString, double> &cell_area, string techname)
+	{
+		tech = techname;
+
+	#define X(_name) _name = 0;
+		STAT_NUMERIC_MEMBERS
+	#undef X
+
+		for (auto wire : mod->selected_wires())
+		{
+			if (wire->name[0] == '\\') {
+				num_pub_wires++;
+				num_pub_wire_bits += wire->width;
+			}
+
+			num_wires++;
+			num_wire_bits += wire->width;
+		}
+
+		for (auto &it : mod->memories) {
+			if (!design->selected(mod, it.second))
+				continue;
+			num_memories++;
+			num_memory_bits += it.second->width * it.second->size;
+		}
+
+		for (auto cell : mod->selected_cells())
+		{
+			RTLIL::IdString cell_type = cell->type;
+
+			if (width_mode)
+			{
+				if (cell_type.in(ID($not), ID($pos), ID($neg),
+						ID($logic_not), ID($logic_and), ID($logic_or),
+						ID($reduce_and), ID($reduce_or), ID($reduce_xor), ID($reduce_xnor), ID($reduce_bool),
+						ID($lut), ID($and), ID($or), ID($xor), ID($xnor),
+						ID($shl), ID($shr), ID($sshl), ID($sshr), ID($shift), ID($shiftx),
+						ID($lt), ID($le), ID($eq), ID($ne), ID($eqx), ID($nex), ID($ge), ID($gt),
+						ID($add), ID($sub), ID($mul), ID($div), ID($mod), ID($divfloor), ID($modfloor), ID($pow), ID($alu))) {
+					int width_a = cell->hasPort(ID::A) ? GetSize(cell->getPort(ID::A)) : 0;
+					int width_b = cell->hasPort(ID::B) ? GetSize(cell->getPort(ID::B)) : 0;
+					int width_y = cell->hasPort(ID::Y) ? GetSize(cell->getPort(ID::Y)) : 0;
+					cell_type = stringf("%s_%d", cell_type.c_str(), max<int>({width_a, width_b, width_y}));
+				}
+				else if (cell_type.in(ID($mux), ID($pmux)))
+					cell_type = stringf("%s_%d", cell_type.c_str(), GetSize(cell->getPort(ID::Y)));
+				else if (cell_type.in(ID($sr), ID($dff), ID($dffsr), ID($adff), ID($dlatch), ID($dlatchsr)))
+					cell_type = stringf("%s_%d", cell_type.c_str(), GetSize(cell->getPort(ID::Q)));
+			}
+
+			if (!cell_area.empty()) {
+				if (cell_area.count(cell_type))
+					area += cell_area.at(cell_type);
+				else
+					unknown_cell_area.insert(cell_type);
+			}
+
+			num_cells++;
+			num_cells_by_type[cell_type]++;
+		}
+
+		for (auto &it : mod->processes) {
+			if (!design->selected(mod, it.second))
+				continue;
+			num_processes++;
+		}
+	}
+
+	void log_data(RTLIL::IdString mod_name, bool top_mod)
```

Comment:
This is an implementation detail of the `stat` pass and shouldn't be a part of global Yosys namespace. This change would make it harder to introduce the `stat -json` option proposed on IRC, for example.

I'm also not really convinced that using `statdata_t` wholesale is the right way to go here.

Reply:
Well, there does need to be some way to measure the size of the module for the `-s` option.  But maybe this is controversial and problematic enough (and frankly unimportant to me) to just remove the `-s` option entirely.

Reply:
To be more clear, I think it's reasonable to have *some* cost function in Yosys core; I just don't it is reasonable to pul the implementation details of the `stat` pass as-is. If it's not important for you I would say ditch it, but I would also be happy to brainstorm a nicer way to do it than the current approach.

Reply:
I think the target size idea is nice--hence my implementing it--but anyway I haven't had a clear idea how to work with a module so partitioned.  A lack of control flow in yosys scripts means there is no scriptable way to detect the number of created partitions and iterate over them (or whatever one needs to do with the created partitions).  And, in a REPL session, the user can run `stat` themselves and easily calculate the desired number of partitions.  So maybe it's not really so useful anyway.

Reply:
Let's ditch it for now; we can always bring it back later.